include/drivers/pio.h

```diff
@@ -13,13 +13,26 @@
 #define PIO_OPS(dev) ((struct pio_driver_ops *)((dev)->drv->ops))
 
 struct pio_driver_ops {
+	bool (*get_port_status)(uint8_t pin);
 	bool (*read_pin)(struct device *dev, uint8_t pin);
 	int  (*set_mode)(struct device *dev, uint8_t pin, uint8_t mode);
 	int  (*write_pin)(struct device *dev, uint8_t pin, bool val);
 };
 
 /**
- * Read the value of a pin.
+ * Checks the status of an I/O port.
```

Comment:
Please explain more what this is doing. An I/O port contains several independent pins, which can be used for different things by different drivers at the same time.

If this is in response to the suggestion I made about keeping a bitmap in the driver data about which ports are available, I meant that in `r_pio->drvdata`, you would have a bitmap of which ports were *physically implemented*. The purpose of that is to give an appropriate error message if a driver tries to use a port that *doesn't exist*. For example, in `r_pio`, only the first port exists, so you would have `.drvdata = BIT(0),` in `devices.c`. Then, when some driver called e.g. `pio_write_pin(&r_pio, SUNXI_PIO_PIN(1, 0), true)`, you could report that no such pin exists (return `ENODEV`). This bitmap of implemented ports doesn't change at runtime, and isn't part of the PIO class API.

Reply:
This was a misunderstanding of our discussion on my part. I have changed the strategy to check for _physically implemented_ pins. Please provide feedback on the new implementation.